docs/rfcs/700-config-features-and-defaultpairs-via-connection-string.md

```diff
@@ -0,0 +1,225 @@
+
+```go
+// DefaultStoragePairs is default pairs for specific action
+type DefaultStoragePairs struct {
+	Create            []Pair
+	Delete            []Pair
+	List              []Pair
+	Metadata          []Pair
+	Read              []Pair
+	Stat              []Pair
+	Write             []Pair
+}
+```
+
+A possible format that follows the struct in connection string for default storage paris may be:
+
+ `default_storage_pairs={write:[storage_class=STANDARD_IA,key2=value],read:[]}`
+
+Obviously, It's complex for users and hard to pass in a valid string. In addition, it either exposes the internal details to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers).
+
+## Proposal
+
+I propose to support config features and default pairs in plain string format in connection string.
+
+### Features
+
+The format of features pairs in connection string is:
+
+`enable_key=value`
+
+- The whole key of the pair SHOULD have the prefix `enable_`.
+- `key` is the pair name defined in `features.toml` and supported in service, and the format SHOULD be exactly the same.
+- `value` SHOULD be `true` or `false`.
+
+So a valid connection string containing features could be:
+
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&enable_loose_pair=false&enable_virtual_dir=true`
+
+The connection string above is equivalent to:
+
+```go
+store, err := s3.NewStorage(
+	ps.WithCredential("hmac:xxxx:xxxx"),
+	ps.WithName("bucket_name"),
+	ps.WithWorkDir("/prefix"),
+	s3.WithStorageFeatures(s3.StorageFeaturs{LoosePair:  true,
+		LoosePair:  true,
+		VirtualDir: true,
+    })
+)
+```
+
+### DefaultPairs
+
+The format of default pair in connection string is:
+
+`default_key=value`
+
+- The whole key of the pair SHOULD have the prefix `default_`.
+- `key` is the pair name defined in `toml` and the format SHOULD be exactly the same.
+- Operation has a different set of default pairs. Pairs in operation pair list will share the same default value if have the same name .
+
+So a valid connection string containing default pairs could be:
+
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&default_server_side_encryption=xxxx&default_strorage_class=xxxx`
+
+Also, we generate the following function for default pairs to support configure the default pair during initialization in the way of `WithXxx()`:
+
+Take `default_storage_class` for example:
+
+```go
+func WithDefaultStorageClass(v string) Pair {
+	reutrn Pair{
+		Key:   "default_storage_class",
+		Value: v,
+	}
+}
+```
+
+The connection string above is equivalent to:
+
+```go
+store, err := s3.NewStorage(
+	ps.WithCredential("hmac:xxxx:xxxx"),
+	ps.WithName("bucket_name"),
+	ps.WithWorkDir("/prefix"),
+    }), 
+    s3.WithDefaultServiceSideEncryption("xxxx"),
+    s3.WithDefaultStorageClass("xxxx"),
+)
+
+```
+
+### Implementation
+
+#### Features
+
+`*Features` are types defined in service.
+
+**Feature Pairs Registry**

```

Comment:
We already have features field in `service.toml`:

```toml
features = ["virtual_dir"]
```

Maybe we can reuse it?

Template that uses features can be found here: https://github.com/beyondstorage/go-storage/blob/master/cmd/definitions/tmpl/service.tmpl#L135-L146

Reply:
Does it mean we can insert `map["virtual_dir"] = "bool"` into `pairMap` instead of generate `*FeaturesPairMap`?

Reply:
https://github.com/beyondstorage/go-storage/blob/master/cmd/definitions/type.go#L283

Maybe we can append into New functions `Optional` directly?

Reply:
Got it.

Reply:
Ok, I got it, we are discussing different things.

There are two things we need to support:

- `DefaultStorageClass` pairs
- `default_storage_class` in connection string.

For the first, we can add pairs into `Optional`, and for the other, we can add into `pairMap`.

Finally, we can reduce the usage of `Feature Pairs Registry` or `Default Pairs Registry`.